TESTING.md

```diff
@@ -0,0 +1,114 @@
+
+**Running these tests can cost you money. They execute against
+paid-for Google services.** They can also produce significant chunks of
+network traffic - particularly some of the integration tests. You're
+not recommended to run them against a mobile internet connection
+unless you're sure you know what you're doing.
+
+**These tests could destroy data in the projects they run in, or
+leave test data afterwards.** Most tests create data in a
+well-partitioned way and clean up after themselves, but there could
+be a bug in a test which wipes all the data in the project. It is
+*very strongly* recommended that you only run tests against projects
+which do not contain any production data or services.
+
+Running tests
+=============
+
+The `build.sh` script will run unit tests, and
+`runintegrationtests.sh` will run integration tests and snippets.
+(TODO: smoke tests too...) Consult the scripts for details on how to
+run.
+
+Additionally, you can just use `dotnet test` in any test project,
+optionally specifying a test filter and/or target framework.
+
+Common environment variables
+============================
+
+To run the integration/smoke tests or snippets, set the following
+environment variables:
+
+- `TEST_PROJECT`: The name of the Google Cloud Project to test against
+- `GOOGLE_APPLICATION_CREDENTIALS`: The absolute path to a JSON file
+  containing service credentials for the test project.
+
+Per API configuration
+=====================
+
+Some APIs require a little more setup. If you're not testing an API
+listed below, the instructions above should be all you need.
+
+Google.Cloud.Bigtable.V2
+------------------------
+
+The Bigtable integration tests can be run against the production
+service or the emulator.
+
+Running in production requires:
+
+- A Bigtable instance, preferably of type Development, created within the
+  project and identified by the `BIGTABLE_TEST_INSTANCE` environment variable
+- The account identified in your service account JSON file to have read/write
+  access to the project, including the IAM permissions in the Bigtable Administrator role
+
+Running against the emulator requires:
+
+- A running emulator on the local machine, started by using the
+  `gcloud beta emulators bigtable start` command
+- The `BIGTABLE_EMULATOR_HOST` environment variable set to `localhost:8086`
+
+Currently there are no runnable snippets, but when they are added, they will
+have the same requirements.
+
+Google.Cloud.Firestore
+----------------------
+
+Firestore cannot currently operate on a Google Cloud project that also has a
+Datastore database. You will need a second project:
+
+- Create a project on the [Firebase Console](https://console.firebase.google.com/u/0/)
+- Enable Firestore on that project (within the Firebase Console)
+- Set the `FIRESTORE_TEST_PROJECT` environment variable with that
+  project ID
+- Add the account identified in your service account JSON file as an
```

Comment:
or run these tests by themselves and change the GOOGLE_APPLICATION_CREDENTIALS?

Reply:
While that would work, it means you can't run `runintegrationtests.sh`. I think this is a simpler approach.

Reply:
I think we should just have it as an options.  This is what I do but I also run all of these tests on semi regularly.  I'm guessing most people who are not us will run one set of tests at a time.   This way users don't have to mess around with switching permissions and adding service accounts

Reply:
Done.